### ISSUE TITLE
Displaced Diacritics

### ISSUE THREAD

**Issue by smeikx** [alacritty/alacritty#1420]
occurs on macOS 10.13.5

# Basic Problem

Using `ls`, file names with diacritics in their name aren’t displayed correctly:
<img width="184" alt="shifted diacritics" src="https://user-images.githubusercontent.com/6908653/42373769-338fae76-8116-11e8-9536-06b1e13a1736.png">

However, this **only** occurs when a file has been (re)named with Finder. 

If I rename the corresponding files via `mv` to contain diacritics they are displayed correctly.
<img width="193" alt="correct" src="https://user-images.githubusercontent.com/6908653/42374809-a6fecf74-8119-11e8-8c38-f8cb4e72b1d9.png">

# Further Complications

**However**, I can’t use `mv` to just overwrite the current name:
<img width="365" alt="renaming" src="https://user-images.githubusercontent.com/6908653/42375422-cd6e0db2-811b-11e8-8c75-941117a10c44.png">

Here, I used `zsh`’s autocompletion, which renders the names correctly; typing the names by hand doesn’t work either. 

In order to make it work I *first* renamed the file to something else, and *then* renamed it back to the originally desired name. 

---

# Additional Info

I don’t know if this is relevant, but when I copy the file names in Finder and paste them in Alacritty it looks like this:
`a<0308>u<0308>o<0308>e<0301>c<0327>n<0303>`

Also, if I copy the incorrectly rendered file names from Alacritty and paste them in Finder (or elsewhere, even in other terminal emulators) they are rendered correctly again.

**Comment by chrisduerr** [comment]
I think this is related to #1317. I'd assume that the origin of the differences is that one time the diacritics are baked into the glyph and the other time they are glyph modifiers. You can see that there is an "empty" space after each cell, this indicates that there is a zero-width character printed there.

**Comment by Mouvedia** [comment]
iTerm on the top
Alacritty on the bottom

![iTermVSalacritty](https://user-images.githubusercontent.com/208817/80318296-7e1af300-8809-11ea-8c98-f08bcb990677.png)

Is that the same bug?

**Comment by kchibisov** [comment]
The font is also different, make it use the same font. and retest, also make sure your locale is the same in both terminals. It seems like it's not.

**Comment by Mouvedia** [comment]
@kchibisov you are right. I am a victim of #2837.
Here's what I did to fix it:

```yml
env:
  LANG: fr_FR.UTF-8
# …
shell:
  program: /usr/local/bin/bash
```

Thanks!

**Comment by chrisduerr** [comment]
The original issue is https://github.com/alacritty/alacritty/issues/2566.

**Comment by Mouvedia** [comment]
@chrisduerr I don't think that's a duplicate.

version: 0.4.1 (4d9253b)
launch: command line
bash: 5.0.0(1)-release

![proof](https://user-images.githubusercontent.com/208817/80416064-cdc5f100-88d3-11ea-86f4-c3477a42d201.png)

**Comment by chrisduerr** [comment]
Didn't you just say that setting `LANG` in Alacritty config fixes it?

**Comment by Mouvedia** [comment]
In my [first comment](https://github.com/alacritty/alacritty/issues/1420#issuecomment-619617037), I was asking if this was the same bug. It wasn't, I just had to set my locale and shell correctly.

As you can see in my second screenshot, his issue is not fixed.

**Comment by chrisduerr** [comment]
@Mouvedia Gotcha, I was looking at the diatrics in the shell, not in the `ls` output.

Though it does look like that was fixed. Are those two separate diatrics, or is the text the same?

**Comment by Mouvedia** [comment]
![téléchargement](https://user-images.githubusercontent.com/208817/80418825-4c249200-88d8-11ea-92d7-063591942612.png)

It should be _téléchargement_.

**Comment by chrisduerr** [comment]
I'm aware, but in your shell working directory it seems to be working correctly. It also seems to work in the `cd` command.

Also is this just an issue with `gls`, or does `ls` also not work?

**Comment by Mouvedia** [comment]
> Also is this just an issue with gls, or does ls also not work?

It doesn't work either with `ls`. I was trying to make a point by using `gls`.

**Comment by chrisduerr** [comment]
If you switch it from `LANG: fr_FR.UTF-8` to `LC_CTYPE: UTF-8`, does that change anything?

**Comment by Mouvedia** [comment]
> If you switch it from LANG: fr_FR.UTF-8 to LC_CTYPE: UTF-8, does that change anything?

It doesn't.

**Comment by kchibisov** [comment]
Could you try different font @Mouvedia ? I guess it's a font related.

**Comment by Mouvedia** [comment]
It's not.

![obviously](https://user-images.githubusercontent.com/208817/80462415-f206e980-8936-11ea-8d3d-fadccad2b009.png)

**Comment by chrisduerr** [comment]
If it was font related, it seems like it should work either both in shell and ls, or in neither.

This seems to be using the same font, but it's displayed differently.

Could you copy the working and non-working strings and inspect if they are byte-for-byte the same?

**Comment by Mouvedia** [comment]
> Could you copy the working and non-working strings and inspect if they are byte-for-byte the same?

Already answered in [my previous comment](https://github.com/alacritty/alacritty/issues/1420#issuecomment-620224893):

Téléchargements vs Te�M-^Ale�M-^Achargements
It's not the same.

**Comment by chrisduerr** [comment]
You acquired both of those by copying it to the clipboard using Alacritty, right? I've asked because your previous comment was talking about using cat, which would be less reliable for checking what is actually stored inside of Alacritty.

**Comment by Mouvedia** [comment]
**unicode code points**

"éléc" :  233, 108, 233, 99
"éléc" : 101, 769, 108, 101, 769, 99 (the one failing)

**Comment by chrisduerr** [comment]
Huh, looks like your shell might perform some kind of unicode normalization, thanks.

Then I'd have to agree with @kchibisov though, must be font related.

**Comment by pqwy** [comment]
It's not font-related. I think that Alacritty doesn't grok [combining characters](https://en.wikipedia.org/wiki/Combining_character).

In a Unicode stream, you often have a [base character](https://www.unicode.org/glossary/#base_character) (like `e`) followed by [combining characters](https://www.unicode.org/glossary/#combining_character) (like [U+0301](https://www.fileformat.info/info/unicode/char/0301/index.htm)), forming a [combining character sequence](https://www.unicode.org/glossary/#combining_character_sequence). The idea is to draw combining characters over the *preceding* base character.

This is the case with the second string. The fact that some of these combining character sequences correspond to individual [precomposed characters](https://www.unicode.org/glossary/#precomposed_character), like [U+00E9](https://www.fileformat.info/info/unicode/char/00e9/index.htm) in the first téléchargement, is unrelated. Both forms are equally valid.

As far as I can tell, Alacritty always advances the cursor after drawing a base character. It then draws the combining character one cell to the right, notices that it's zero-width, keeps the cursor, and draws the next character in the same place, resulting in an off-by-one.

For good measure, here's a comparison of Alacritty, Gnome Terminal, Xterm and Urxvt on a [sample file](https://github.com/alacritty/alacritty/files/4555703/sample.txt) that combines characters `a`, `b`, and `c` with combining marks U+0300 to U+036F. All were using Adobe Source Code Pro.

![chars-alacritty](https://user-images.githubusercontent.com/325535/80663814-f2b19400-8a8c-11ea-82d3-31a2b40394c1.png)
![chars-gnome-terminal](https://user-images.githubusercontent.com/325535/80663817-f6ddb180-8a8c-11ea-91dd-77164773f792.png)
![chars-xterm](https://user-images.githubusercontent.com/325535/80663821-fb09cf00-8a8c-11ea-9a17-b65f5582c751.png)
![chars-urxvt](https://user-images.githubusercontent.com/325535/80663826-fe04bf80-8a8c-11ea-9da0-39813c503472.png)

**Comment by kchibisov** [comment]
@pqwy It's font related, your test works just fine. Firefox also fails on your test for example with DejaVu Sans Mono, but can handle it just fine with Fira Mono.

![picture](https://user-images.githubusercontent.com/27620401/80664538-b176af80-8a9f-11ea-88e1-e6aab77bc1b7.png)

**Comment by pqwy** [comment]
Oh wow, it does work with Fira!

Sorry for the noise, then.

**Comment by kchibisov** [comment]
We do explicitly handle zerowidth chars and their positions, the problem is that some fonts are using different offsets for them, which doesn't make any sense. Why so? Because they use the exact same offsets as for normal characters, which is just strange and doesn't make any sense at all. So the only solution is to workaround broken fonts, which we're against of doing, because fonts should be fixed, since we just honor what they are telling us to do.

**Comment by Mouvedia** [comment]
If that's really the case then [JetBrains Mono](https://www.jetbrains.com/lp/mono/) is broken.

**Comment by kchibisov** [comment]
It seems like 1.0.6 is still broken, since Firefox fails on mentioned `sample.txt` exactly like alacritty.

**Comment by chrisduerr** [comment]
Closing since I don't think there's anything we should do here.

If you have reason to believe that this is not font metric related @smeikx, please let me know and I'll re-open.

**Comment by kchibisov** [comment]
fyi, upstream jetbrains mono issue. https://github.com/JetBrains/JetBrainsMono/issues/228 . They don't have those symbols at all.

**Comment by pqwy** [comment]
I understand not wanting to support brokenness. But I think you should be clearer that this is a *hard minority* position.

Here's several programs and their handling of the issue. It's clear that the choice of font has an impact; it's equally clear that the choice of program has an impact too.

Firefox actually does quite well with DejaVu Sans Mono. Xterm manages perfect rendering in all cases, Gnome Terminal almost does, while Alacritty stands out with by far the worst overall rendering.

Note that no terminal around does proper text shaping (what Harfbuzz does), as that would destroy the concept of a cell. (You can confirm this by pasting a snippet of Arabic.) Yet, mature terminals handle combining characters, regardless of the font.

If I had to guess, I'd guess that the others are using additional information available purely *at the Unicode level* to detect combining sequences. This is my point: combining characters are not purely a *presentation* (i.e. font) thing. These sequences are well-defined and detectable units in the character stream itself, but just checking if a character is zero-width is not enough.

Slightly less confusing [sample.txt](https://github.com/alacritty/alacritty/files/4560749/sample.txt).

Left to right: DejaVu Sans Mono, Fira Mono, Source Code Pro.

![fx](https://user-images.githubusercontent.com/325535/80750332-ca26aa00-8b1f-11ea-868d-fdedccf02849.png)

![xterm](https://user-images.githubusercontent.com/325535/80750341-cdba3100-8b1f-11ea-8f88-e09b62e62535.png)

![gnome-terminal](https://user-images.githubusercontent.com/325535/80750351-d1e64e80-8b1f-11ea-9ee1-039502c98266.png)

![alacritty](https://user-images.githubusercontent.com/325535/80750360-d6126c00-8b1f-11ea-973e-f12ceedbfc79.png)

**Comment by kchibisov** [comment]
>If I had to guess, I'd guess that the others are using additional information available purely at the Unicode level to detect combining sequences. This is my point: combining characters are not purely a presentation (i.e. font) thing. These sequences are well-defined and detectable units in the character stream itself, but just checking if a character is zero-width is not enough.

The guy who is creating JetBrains Mono told explicitly that font should have negative metrics to make it work(which DejaVu Doesn't have), I kind of trust him here, and knowing that DejaVu Sans Mono doesn't work for me at all in Firefox on a lot of cases. You can patch alacritty and keep your patch around, I don't want add a workaround based on font width information, if you can prove that we're missing some font information, go ahead, but I don't want to waste my time looking at this issue one more time, since last time I've done it, it pointed out on wrong bitmap metrics for such glyphs.

**Comment by Mouvedia** [comment]
#2620 is related

**Comment by smeikx** [comment]
I’m not sure I’m qualified to decide the fate of this issue, @chrisduerr. 
I agree with @pqwy, though: If many/most other emulators (on Mac, notably Terminal.app and iTerm) handle combining characters well, Alacritty should also be able to do so.

Regarding the question if the fonts are broken, I’m consulting type designers/engineers for further input.

**Comment by djrrb** [comment]
Hi all, I’m the developer of the [Input](http://input.fontbureau.com) font used in the screenshot in the original post. I’m not an expert on combining accents but @smeikx asked me to weigh in.

> @kchibisov: We do explicitly handle zerowidth chars and their positions, the problem is that some fonts are using different offsets for them, which doesn't make any sense. Why so? 

As support for combining marks has change, so have strategies for making them in fonts.

1) Some fonts have the zero-width combining accents offset to the left so the mark will kinda float over/under a lowercase letter that precedes it by default. It’s a blunt solution, but the benefit is that it works even if the mark positioning [GPOS table](https://docs.microsoft.com/en-us/typography/opentype/spec/gpos) adjustments are ignored by the text renderer.

2) Other fonts have the combining marks centered on their zero width, and rely exclusively on GPOS table to make them look good with the preceding letter.

3) And other fonts (like my typeface Input) center the combining accents, but do not contain any mark positioning GPOS rules. These may have been developed before combining accents were well-supported and have the expectation that users will be using and seeing the precomposed characters instead of the combining marks ([É](https://www.fileformat.info/info/unicode/char/00c9/index.htm) instead of [E](https://www.fileformat.info/info/unicode/char/0045/index.htm) + [combining  ́](https://www.fileformat.info/info/unicode/char/0301/index.htm)).

With that context, it seems like there are two things in question here:

**1) Are the fonts broken?**

Input and fonts like it would doubtlessly be enhanced with mark positioning rules to better place them over the preceding character...I’m happy to add that feature to the to-do list.

But since these fonts support precomposed accents and those are commonly employed by text editors, it is hard for me to think of the font as “broken.” Input is six-year-old software and I think this is the first time I’ve been asked to speak to this issue.

**2) Should a text editor mess with the character stream to optimize the presentation of these combinations?** 

I’m not qualified to answer this, but it seems like a text editor can take cues from how the OS handles this sort of thing in order to have the best support for the most fonts. 

As an example, here’s what happens in Input when I place a letter followed by the combining acute in TextEdit. When the precomposed equivalent exists, it is displayed and everything looks good. When no precomposed character exists, it doesn’t look great but it’s better than nothing:

<img width="585" alt="Screen Shot 2020-05-03 at 10 09 15 AM" src="https://user-images.githubusercontent.com/6578861/80916356-265b1b00-8d26-11ea-95ce-658d5aa4fd62.png">

Apologies for the lengthy post, and hope something in here is helpful.

**Comment by chrisduerr** [comment]
@djrrb Sounds like it is indeed an issue that should be fixed in the font, thanks for weighing in.

**Comment by smeikx** [comment]
> Sounds like it is indeed an issue that should be fixed in the font, thanks for weighing in.

Considering @djrrb’s input (thanks again) and that most major terminal emulators support some kind of normalisation, I disagree. 

----

Also, I did some further testing with the following string: ĄįąęÄÖÜËäöüëéèÉ

I used Monaco and Fira Mono. Both obviously struggle with uppercase letters. Fira can’t deal with the ogonek at all (in the first four characters), Monaco doesn’t handle it well either.

Fira Mono:
![Fira Mono](https://user-images.githubusercontent.com/6908653/80921218-58e03480-8d75-11ea-8ccd-6e88e3f604b0.png)

Monaco:
![Monaco](https://user-images.githubusercontent.com/6908653/80921223-5da4e880-8d75-11ea-8b4a-5ed3cf7685e3.png)

Terminal.app with Input for reference:
![Input in System Terminal](https://user-images.githubusercontent.com/6908653/80921222-5d0c5200-8d75-11ea-84a6-a74618035239.png)

Only relying on font metrics to correctly place combining diacritics will not work well for uppercase letters. This is something that can’t be *‘fixed’* in fonts without distorting certain proportions, it needs to be supported by the software rendering them.

----

(I hope I didn’t get some terms wrong, this is all quite confusing.)

**Comment by pqwy** [comment]
@smeikx This [particular string](https://github.com/alacritty/alacritty/files/4572746/aiae.txt) has a peculiar property: all the combining sequences correspond to certain precomposed forms.

I.e. [U+0041](https://www.fileformat.info/info/unicode/char/0041/index.htm) followed by [U+0328](https://www.fileformat.info/info/unicode/char/0328/index.htm) canonically maps to [U+0104](https://www.fileformat.info/info/unicode/char/0104/index.htm); [U+0075](https://www.fileformat.info/info/unicode/char/0075/index.htm) followed by [U+0308](https://www.fileformat.info/info/unicode/char/0308/index.htm) canonically maps to [U+00FC](https://www.fileformat.info/info/unicode/char/00fc/index.htm), etc.

Terminal.app and Xterm* will run [Unicode Normalization Algorithm](http://www.unicode.org/reports/tr15) over strings they are about to display, producing the NFC (Canonical Composition) form. In this form, combining sequences that can be, are replaced with precomposed characters.

By the time it comes to fetching glyphs, both work with e.g. U+0104 or U+00FC, and draw nice hand-precomposed characters.

(That is, that's what I _guess_ happened in your case. As NFC is the recommended normalization form for the Web, the `Aiae` string above is served by Github in NFC and was likely converted by your browser. :grinning:)

@chrisduerr This is a bit orthogonal to handling real combining forms like `x⃗`, but running the fast and standard [normalization algorithm](https://github.com/unicode-rs/unicode-normalization) over the output – and converting to NFC – would vastly reduce the number of cases where this even matters. Maybe that is decent middle ground?

[*] Xterm uses a [home-baked approximation](https://github.com/joejulian/xterm/blob/master/charproc.c#L1720-L1727) unsurprisingly.

**Comment by chrisduerr** [comment]
> @chrisduerr This is a bit orthogonal to handling real combining forms like x⃗, but running the fast and standard normalization algorithm over the output – and converting to NFC – would vastly reduce the number of cases where this even matters. Maybe that is decent middle ground?

There were some issues discussing unicode normalization in the past, but I don't really see a reason to do so. Applications are in full control over what should be displayed, so I don't see why we would render anything other than what the application tells us to render. If they want to show the user the normalized string, the application can normalize it.

**Comment by smeikx** [comment]
> Applications are in full control over what should be displayed, so I don't see why we would render anything other than what the application tells us to render. If they want to show the user the normalized string, the application can normalize it.

So, you think that it should be `ls`’ responsibility to do the normalization in order to display file names as they were intended to be displayed? Do I understand correctly, @chrisduerr?

***

I tried GNU `ls` as well, with the same results. Also, I didn’t find a way to activate normalisation, neither in GNU nor in BSD (macOS) `ls`’ `man` page.

***

If I want file names containing diacritics, that have been named with Finder, to be displayed as intended, using a font of my choice, I can only think of two options:

- create an alias for `ls` piping it through `iconv -f utf-8-mac -t utf-8`, losing colours
- rename all affected files (periodically) on the command line, maybe using `find` an `iconv`

These options don’t feel like good *solutions* to me, though.

**Comment by chrisduerr** [comment]
> to display file names as they were intended to be displayed

If they're supposed to be displayed the way they were intended to be displayed, they shouldn't be normalized. Normalization is an attempt to change the way it is displayed. If you want the normalized for to be displayed, you should store the filename as such.

**Comment by Mouvedia** [comment]
@chrisduerr could we go about it the feature request way?
What I mean is, would you be willing to add an option in the yml config to turn on the normalization by default?

Of course this would be a new issue, separate from this wontfix.

**Comment by pqwy** [comment]
> Normalization is an attempt to change the way it is displayed.

Keep in mind that in Unicode, normalization (to any of 4 forms) is considered to preserve _canonical equivalence_—these strings are strictly the same text.

Now that I'm looking into it, so far I can confirm that Xterm, Urxvt, and Gnome Terminal convert to NFC, and it's obvious from the screenshot that Terminal.app does. [W3C](https://www.w3.org/TR/charmod-norm/) has a long recommendation about this, too.

Seems to be well­­-established practice.

**Comment by chrisduerr** [comment]
> What I mean is, would you be willing to add an option in the yml config to turn on the normalization by default?

I don't see a reason to support this.

**Comment by pqwy** [comment]
> I don't see a reason to support this.

This is well-established practice in the terminal world. It is a clear use-case of Unicode normalization.

**Comment by Mouvedia** [comment]
JetBrains Mono was partially fixed in JetBrains/JetBrainsMono@7445d1f0315a3d12c7cc77421593f8854c6794b2
